File: app/modules/vlm-extractor/ios/VlmExtractorModule.swift
import ExpoModulesCore

/// PR 1 scaffolding. All entry points report that the VLM path is not
/// available on this device, which makes `extractFromLabel` fall back to
/// the existing OCR + rules pipeline. MLX + Gemma-3n E2B land in PR 2.
public class VlmExtractorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("VlmExtractor")

    AsyncFunction("isAvailable") { () -> [String: Any] in
      return ["supported": false, "reason": "unsupported_os"]
    }

    AsyncFunction("prepareModel") { () in
      // Intentionally empty; real implementation downloads and memory-maps
      // the MLX checkpoint in PR 2.
    }

    AsyncFunction("extract") { (_ imageUri: String) -> [String: Any] in
      throw Exception(
        name: "VlmExtractorUnavailable",
        description: "VLM extract is not yet implemented on iOS"
      )
    }
  }
}

File: app/modules/vlm-extractor/android/src/main/java/expo/modules/vlmextractor/VlmExtractorModule.kt
package expo.modules.vlmextractor

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * PR 1 scaffolding. All entry points report that the VLM path is not
 * available on this device, which makes extractFromLabel fall back to
 * the existing OCR + rules pipeline. MediaPipe LlmInference + Gemma-3n
 * E2B land in PR 3.
 */
class VlmExtractorModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("VlmExtractor")

    AsyncFunction("isAvailable") {
      mapOf("supported" to false, "reason" to "unsupported_os")
    }

    AsyncFunction("prepareModel") {
      // Intentionally empty; real implementation downloads and loads the
      // MediaPipe .task bundle in PR 3.
    }

    AsyncFunction("extract") { _: String ->
      throw VlmExtractorUnavailableException()
    }
  }
}

private class VlmExtractorUnavailableException :
  CodedException("VLM extract is not yet implemented on Android")
